apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs

```diff
@@ -107,7 +111,17 @@ private static SpannerSettings CreateSpannerSettingsWithVersionHeader()
         /// <param name="logger">The logger to use. May be null, in which case the default logger is used.</param>
         /// <returns>A <see cref="SessionPoolManager"/> with the given options.</returns>
         public static SessionPoolManager Create(SessionPoolOptions options, Logger logger = null) =>
-            new SessionPoolManager(options, logger ?? Logger.DefaultLogger, CreateClientAsync);        
+            new SessionPoolManager(options, DefaultSpannerSettings(), logger ?? Logger.DefaultLogger, CreateClientAsync);
+
+        /// <summary>
+        /// Creates a <see cref="SessionPoolManager"/> with the specified SpannerSettings and options.
+        /// </summary>
+        /// <param name="spannerSettings">The SpannerSettings to use. May be null, in which case the default settings are used.</param>
+        /// <param name="options">The options to use. Must not be null.</param>
+        /// <param name="logger">The logger to use. May be null, in which case the default logger is used.</param>
+        /// <returns>A <see cref="SessionPoolManager"/> with the given options.</returns>
+        public static SessionPoolManager Create(SessionPoolOptions options, SpannerSettings spannerSettings, Logger logger = null) =>
+            new SessionPoolManager(options, spannerSettings ?? DefaultSpannerSettings(), logger ?? Logger.DefaultLogger, CreateClientAsync);
```

Comment:
SpannerSettings already has a logger... I'm wondering whether this overload should *not* take a logger, but use the one from there. (Basically, if you're customizing the settings already and you want a custom logger, put it in the settings.)

If we do still want the logger parameter, I'd suggest using `logger ?? spannerSettings.Logger ?? Logger.DefaultLogger`.

Reply:
(But having it separately in the internal constructor still makes sense, for the case where the settings *aren't* specified explicitly.)

Reply:
That does make sense, __however__: It could possibly break existing applications that are currently calling `SessionPoolManager.Create(options, null)`, as that would then become an ambiguous invocation. So I guess using `logger ?? spannerSettings.Logger ?? Logger.DefaultLogger` would be the safer option.

Reply:
It won't be ambiguous - it will unambiguously use the new overload (which won't have any optional parameters). It's a little odd that a version change would change the selected overload - but the effect (default settings, default logger) would be the same.

It doesn't change binary compatibility as any code *previously* compiled (where it picks the original method) will have the method signature in the IL already, and that method still exists.

Reply:
I think it will be ambiguous, but __only__ for anyone that is calling the current method with an unnecessary `null` for the second argument (the logger):

```csharp
var mgr = SessionPoolManager.Create(new SessionPoolOptions(), null);
```

If we were to remove the `Logger` argument from the new overload, we would have the following method definitions:

```csharp
public static SessionPoolManager Create(SessionPoolOptions options, Logger logger = null) => ...

public static SessionPoolManager Create(SessionPoolOptions options, SpannerSettings spannerSettings) => ...
```

So I don't think there will be very many (if any) who does, but in theory it could cause a compilation error in existing code.

Or did I misunderstand how you wanted to define the new overload without the logger argument?

Reply:
I promise you, it really won't be ambiguous. Overload resolution has a tie-break that means a candidate method that doesn't need the compiler to fill in any optional parameters with default values is "better than" a method that doesn't.
(But don't just take my word for it - try it :)

Those overloads are indeed what I'd be expecting.

Reply:
> (But don't just take my word for it - try it :)

I know that I'm communicating with people with a lot more C# experience than myself here, so to be a little bit more confident I did try it before I stated my initial "I think it will break..." :-)

But apparently not all compilers (or IDEs) agree in this case:

![Screenshot from 2021-07-21 13-50-16](https://user-images.githubusercontent.com/1196707/126484465-eed340a0-b77b-4ca0-b36d-f1826d503bb0.png)

Reply:
Interesting. Will try it in VS by pulling from your branch first...
(I'm very happy to be proved wrong... it's going to be more awkward to decide what to do if this turns out to be a Rider bug.)

Reply:
VS agrees with you. Off for lunch now, but will investigate further afterwards.

For reference, this is what I expect to apply: https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#12643-better-function-member

In particular:

> If neither function member was found to be better, and all parameters of Mᵥ have a corresponding argument whereas default arguments need to be substituted for at least one optional parameter in Mₓ, then Mᵥ is better than Mₓ. Otherwise, no function member is better.

I need to work out why that doesn't apply here...
(And obviously we need to work out what to do in terms of the methods... having an extra unnatural parameter doesn't feel ideal. I wonder whether we should rename the one that accepts settings instead?)

Reply:
Oh, doh - I'm thinking of the rule the wrong way round ignore me, and I'm very glad you tested it. That rule is what makes the logger version "win" if you had an overload of 

```csharp
public static SessionPoolManager Create(SessionPoolOptions options, Logger logger = null, SpannerSettings spannerSettings = null)
```

Reply:
I've removed the `Logger` argument from the overload and renamed the method to `CreateWithSettings` (name open for discussion). I agree that the additional `Logger` argument only makes it more complex.

Reply:
Fab, thanks. Will look after lunch. Sorry again about the incorrect assertion around ambiguity...

Reply:
Ah, and the one I tested wasn't ambiguous because string is a subtype of object, and that's why the method with the string was chosen.